### PR TITLE
Engine API interop: accept op-node's bare-number minBaseFee, store the eth-genesis timestamp in the millisecond domain

### PR DESCRIPTION
**Pull request by kyonRay**
## What

Two Engine API interop fixes found by running a real **op-node v1.19.3** (official docker image, sequencer mode) against a node built at release-3.18.0 head with `[op_engine_rpc]` enabled. Both were hard blockers: with them in place the sequencing loop reaches `Sequencer sealed block number=1` end to end; without them op-node cannot issue a single successful forkchoiceUpdated with attributes.

### 1. Accept a bare JSON number for `payloadAttributes.minBaseFee`

op-node declares `MinBaseFee *uint64` with no hexutil wrapper (`op-service/eth/types.go:523`), so the wire carries `"minBaseFee": 0` — a bare JSON number, unlike every other quantity field it sends (those are hex `Uint64Quantity` strings). Our parser accepted only hex quantity strings and answered -32602, which op-node classifies as a temporary engine error and retries forever.

The new `parseQuantityOrUInt64Number` admits unsigned JSON integers via the jsoncpp type tag. Two deliberate choices: it never goes through `asString()` (jsoncpp stringifies numbers, so `10` would re-parse as hex 16 — silent value forgery), and it does not use `isUInt64()` as the gate (that predicate admits integral doubles like `10.0`). Only the minBaseFee call site uses the new helper; `timestamp`/`gasLimit` stay hex-string-only, pinned by a dedicated test so the accepted surface cannot silently widen.

### 2. Store the eth-genesis timestamp in the internal millisecond domain

Since #5420, `applyEthGenesisHeader` copied the artifact's timestamp (seconds, the Ethereum header domain) verbatim into the internal millisecond-domain header field. That left one seconds-valued outlier inside a millisecond domain, and every consumer needed a special case for B0: the web3 block view divided it by 1000 and reported January 1970 (op-node then refuses to sequence: `cannot build L2 block ... before L1 origin`, and `gen_rollup_config.py` derived a wrong `l2_time`), and the EVM `TIMESTAMP` path (`HostContext.cpp`) had the same latent bug.

**Timestamp domain rule (maintainer decision), which this PR implements:** the internal `BlockHeader.timestamp` carries **milliseconds for every header version**; conversion to the Ethereum seconds domain happens only at the boundaries — RLP encoding (unconditional ÷1000), RLP/wire decoding (unconditional ×1000, with an `INT64_MAX/1000` bound rejecting wire values that would overflow, covered by a red test), EVM execution, and RPC output. The existing unconditional ÷1000 in the web3 block view and in the EVM `TIMESTAMP` path are already correct under this rule — no version-keyed branches anywhere.

This deliberately supersedes the per-version placement introduced by #5434 (`rlpEncode` divided only for `NON_ETH` and passed ETH-version headers through, implying ETH-version headers hold seconds internally). The rationale is to avoid version-keyed if/else spreading through every consumer. Concretely on top of #5434: `rlpEncode` now always divides (and its sub-second rejection applies to every version — ETH-version headers are only ever produced from whole-second sources, verified by caller inventory); `rlpDecode` multiplies (the overflow guard sits here so all three RLP→header paths — `decodeTarsHeader`, `toTarsHeader`, `toEthBlockHeader` — share one conversion point and decode→re-encode round-trips stay canonical); `toTarsHeader`/`decodeTarsHeader` pass through what is now already milliseconds. `validateHeader` additionally rejects sub-second timestamps. `applyEthGenesisHeader` stores `seconds * 1000`, and the `[eth_genesis_header].timestamp` config bound tightens to `INT64_MAX/1000` so the multiplication cannot overflow. `ms = s * 1000` round-trips losslessly, so the genesis hash is byte-for-byte unchanged — the offline python reference hash in `GenesisEthHeaderTest` and the golden mainnet Cancun header hash + RLP bytes both pass unmodified.

**Breaking note (unreleased feature only):** chains built with an `[eth_genesis_header]` before this change store B0 in seconds and will fail the genesis-hash consistency check on restart; rebuild the datadir. Legacy chains are unaffected (their B0 never sets `ethBlockVersion` and is already in milliseconds; NON_ETH headers never enter the RLP-hash path, `BlockHeaderImpl.cpp` keeps them on the Tars hash branch).

## Verification

- minBaseFee: 6 new unit cases (bare `0` -> 0, bare `10` -> **10, not 16**, `UINT64_MAX` boundary, hex string still accepted; `-1` / `1.5` / `"notahex"` / `[]` / `{}` rejected with -32602; timestamp/gasLimit bare numbers still rejected). The mock CL now sends `minBaseFee` as a bare number, matching the real op-node wire shape.
- Timestamp domain (on top of the rebased base 63e131920): `test-bcos-rlp-protocol` 25/25 (golden mainnet header hash `95d7f597…` and RLP bytes unchanged; new sub-second-rejection and overflow-rejection cases), `test-bcos-ledger` green incl. all 7 `GenesisEthHeaderTest` cases (python reference `b153f41d…` unchanged) — the only red suite is `ComputeTrieRootSuite`, a pre-existing Debug-only assert introduced with the base itself (zero MPT files in this diff); `test-bcos-rpc`, `test-bcos-tars-protocol`, `test-bcos-codec`, `test-bcos-framework`, `test-transaction-scheduler`, `test-bcos-tool` all green (the latter five cover the tests the base PR added).
- Live run: op-node v1.19.3 connected **directly** to port 8551 (the first smoke needed a JSON-rewriting proxy to get past minBaseFee). Zero -32602, block 1 sealed and inserted, `eth_getBlockByNumber("0x0")` returns `0x689d5c00`, regenerated rollup.json carries `l2_time: 1755143168`. The session then stops at the next known gap (committed engine blocks exposed no transactions over RPC — addressed in #5445), with no new errors before it.

## Notes for follow-up parts of the OP series

- New header producers must write **milliseconds** into `BlockHeader.timestamp` (the Engine API boundary already converts attribute seconds ×1000); consumers of `EthBlockHeader::data().timestamp` now receive milliseconds and must not multiply again.
- `computeHash`/`rlpEncode` throw on sub-second millisecond values — a loud failure rather than a silently wrong hash if a legacy `utcTime()` header ever reaches the eth-hash path.

### CODE REVIEW

File: bcos-rlp-protocol/bcos-rlp-protocol/EthBlockHeader.cpp [hunk 1]
```diff
@@ -320,7 +332,10 @@ EthBlockHeader::EthBlockHeader(const bcos::protocol::BlockHeader& _header)
     m_data.gasLimit = _header.gasLimit();
     m_data.gasUsed = _header.gasUsed();
     m_data.number = _header.number();
-    m_data.timestamp = _header.timestamp();
+    // Domain bridge: internal BlockHeader timestamps are milliseconds; the
+    // Ethereum wire/RLP domain is seconds. Convert exactly here so the
+    // encoding side always sees seconds.
+    m_data.timestamp = _header.timestamp() / 1000;
```
Comment (morebtcg) [inline bcos-rlp-protocol/bcos-rlp-protocol/EthBlockHeader.cpp]: This `/1000` silently truncates any sub-second remainder: an internal eth header carrying e.g. `1700000000123` ms would be hashed as `1700000000` s, and a decode round-trip through `toTarsHeader` would not reproduce the header that was hashed. Today every producer lands on whole seconds (Engine boundary x1000, `applyEthGenesisHeader` x1000, decode x1000), so the invariant holds structurally — but nothing enforces it at this choke point. Consider having `validateHeader` (which `calculateRLPHash` runs just before this constructor) reject `timestamp % 1000 != 0`, or at least state the whole-second invariant here, so a future caller can't silently hash a different header than the one it is holding.

File: bcos-rlp-protocol/bcos-rlp-protocol/EthBlockHeader.cpp [hunk 2]
```diff
@@ -60,6 +61,16 @@ bcos::Error::UniquePtr EthBlockHeader::toTarsHeader(
         return err;
     }
 
+    // The wire timestamp is seconds; the internal header stores milliseconds.
+    // Bound it by int64 before the x1000 below (mirroring the Engine boundary
+    // contract) so a hostile wire value cannot trigger signed overflow.
+    constexpr int64_t c_maxWireTimestampSeconds = std::numeric_limits<int64_t>::max() / 1000;
+    if (ethHeader.data().timestamp < 0 || ethHeader.data().timestamp > c_maxWireTimestampSeconds)
```
Comment (morebtcg) [inline bcos-rlp-protocol/bcos-rlp-protocol/EthBlockHeader.cpp]: This guard covers two distinct overflow routes into the x1000 below: wire values >= 2^63 wrap to negative via the `static_cast<int64_t>` in `rlpDecode` (caught by `< 0`), and positive values in `(INT64_MAX/1000, INT64_MAX]` (caught by the upper bound). The new `toTarsHeaderRejectsOverflowingTimestamp` test only exercises the first branch (2^63). Worth adding a case with a positive-but-too-large value — e.g. `INT64_MAX/1000 + 1`, which encodes fine as an RLP scalar and would UB on the x1000 — so the upper-bound branch can't regress silently.

File: bcos-ledger/bcos-ledger/Ledger.cpp [hunk 3]
```diff
@@ -1984,9 +1984,13 @@ static void applyEthGenesisHeader(
...
-    // B0's timestamp is artifact-authoritative and in SECONDS (the Ethereum
-    // header domain): the RLP hash must reproduce the artifact byte-for-byte.
-    header.setTimestamp(ethHeader.m_timestamp);
+    // The artifact carries B0's timestamp in SECONDS (the Ethereum header
+    // domain); internal BlockHeader timestamps are MILLISECONDS everywhere, so
+    // convert on the way in. The RLP bridge (EthBlockHeader) divides by 1000 on
+    // encode, so keccak256(rlp(header)) still reproduces the artifact hash
+    // byte-for-byte. No int64 overflow: artifact seconds (~2^34) x 1000 is far
+    // below INT64_MAX.
+    header.setTimestamp(ethHeader.m_timestamp * 1000);
```
Comment (morebtcg) [inline bcos-ledger/bcos-ledger/Ledger.cpp:2010]: The comment rules out overflow for the current artifact (~2^34 s), but `m_timestamp` is operator-supplied int64 config and `NodeConfig` only bounds it at `INT64_MAX` (NodeConfig.cpp:370). A value >= ~9.2e15 seconds passes config validation and makes `m_timestamp * 1000` signed overflow (UB) here, before the m_hash consistency check gets its chance to fail-fast on the mismatch. Consider tightening the NodeConfig bound to `INT64_MAX / 1000` — mirroring the bridge's `c_maxWireTimestampSeconds` — so a bad artifact is rejected with a clear config error instead of UB.

**Review by morebtcg — COMMENTED** [review]
Reviewed at 816039a. The overall design holds up — I spot-checked the caller-inventory claims in the description and they check out:

- The only production RLP-hash path is `BlockHeaderImpl::calculateHash` -> `EthBlockHeader::calculateRLPHash`, and every producer of an eth-versioned internal header lands on whole-second milliseconds (Engine boundary `engineSecondsToInternalMillis`, `applyEthGenesisHeader` x1000, `toTarsHeader` x1000), so the /1000 encode bridge round-trips losslessly today.
- `HostContext::timestamp()` already divides by 1000 under `feature_evm_timestamp`, so storing B0 in ms does fix the EVM `TIMESTAMP` path for free as claimed.
- `parseQuantityOrUInt64Number` correctly avoids the `asString()` hex-reparse trap by gating on the jsoncpp type tag, and scoping bare-number acceptance to the minBaseFee call site (with the pin test) is the right surface control.
- The new overflow guard in `toTarsHeader` is placed before any field is written to the destination header, so the early return leaves it in the cleared state — good.

Three minor comments inline: one silent-truncation invariant worth enforcing at the bridge, one test-coverage gap on the new overflow guard, and one signed-overflow edge reachable from config input. None of them block the intent of this PR.

**Review by morebtcg — APPROVED** [review]
Approving at 816039a. The three inline comments from my earlier review (whole-second truncation invariant at the RLP encode bridge, missing test for the positive upper-bound branch of the new overflow guard, and the NodeConfig timestamp bound vs. the x1000 in applyEthGenesisHeader) are all minor hardening suggestions — none of them block the two interop fixes, which I verified against the actual call paths (Engine boundary, calculateHash, HostContext, genesis build). Fine to merge; the follow-ups can land separately.

File: bcos-rlp-protocol/bcos-rlp-protocol/EthBlockHeader.cpp (hunk 2 above)
Comment (Shichen-Wu) [inline bcos-rlp-protocol/bcos-rlp-protocol/EthBlockHeader.cpp]: `< 0` 分支依赖 `rlpDecode`（:419 `m_data.timestamp = static_cast<int64_t>(_timestamp)`）把 uint64 wire 值强转 int64 时的 **implementation-defined wrap**。当恶意 RLP header 的 timestamp ≥ 2^63 时，`static_cast<int64_t>(uint64_t)` 按 C++ [conv.integral] 是 implementation-defined（不保证 wrap），而这个 guard 靠「wrap 成负值」来捕获它——若某编译器（或未来标准/UBSan 严格模式）不 wrap，超大值会绕进 `*1000`。三主流编译器目前都 wrap、且此依赖在 base 的 `validateHeader` 就存在，故 P3；但这是 decode 恶意 wire 输入的安全敏感路径，建议在 `rlpDecode` 里显式 `if (_timestamp > INT64_MAX) return error`，让拒绝不依赖未定义行为。（P3）

**Review by Shichen-Wu — COMMENTED** [review]
Round 1（本轮独立评审，在 morebtcg APPROVE 之后）· PR 5442，at `816039a49`（base `dbf353b71`）。10 files / +262 / −15，两个 Engine API 互操作修复 + timestamp 域统一。

**结论：COMMENT。核心逻辑自洽，无新 P0/P1；morebtcg 的 3 条 minor（whole-second 截断不变量、overflow guard 缺 positive upper-bound 测试、NodeConfig bound vs ×1000 overflow）我独立复核后全部认同。我的增量为 1 条 P3 精确性观察 + 若干证伪项。是否 approve 由你决定。**

## 一、独立复核核心逻辑（均自洽）

1. **timestamp 域统一「单一 ms 域 + 桥接转换」**：三处改动方向一致——`applyEthGenesisHeader` ×1000（秒→ms 存储）、`EthBlockHeader(const BlockHeader&)` 构造器 ÷1000（ms→秒编码）、`toTarsHeader` ×1000（秒→ms 解码，带 overflow guard）。我验证了 hash 一致性：`calculateRLPHash` 走构造器 ÷1000 后 `rlpEncode` 直接编码秒；`toTarsHeader` 走 `rlpDecode` 得到秒、`rlpEncode(reencoded)` 也编码秒——两条 hash 路径都编码秒，round-trip 一致。✓
2. **`parseQuantityOrUInt64Number` 类型门正确**：`Json::uintValue || (intValue && asInt64()>=0)` 精确覆盖「非负整数」；负数（`-1` 是 intValue 负值）与 real（`1.5`/`10.0` 是 realValue）都 fall-through 到拒绝；不经过 `asString()` 避免 `10`→hex 16 的静默伪造。且其 hex string 分支与 `parseQuantity`（base 已有的 hex-only reader）**完全同源**（都走 `safeFromQuantity`），minBaseFee 的 hex 形式与其他字段行为一致。✓
3. **overflow guard 位置正确**：`toTarsHeader` 的 guard 在 `header->clear()` 之后、任何字段写入之前，early return 时 header 处于 cleared 状态（morebtcg 已确认）。

## 二、我的新发现（morebtcg 未覆盖）

**Finding 1（P3）——overflow guard 的 `< 0` 分支依赖 `static_cast<int64_t>(uint64_t)` 的 implementation-defined wrap。** 见行内 `EthBlockHeader.cpp:68`。`rlpDecode`（:419）`m_data.timestamp = static_cast<int64_t>(_timestamp)` 把 uint64 wire 值强转 int64；当恶意 wire 值 ≥ 2^63 时，这个转换是 **implementation-defined**（C++ [conv.integral]，非 guaranteed wrap），而 guard 的 `< 0` 分支**依赖**这个 wrap 把超大值变负来捕获。GCC/Clang/MSVC 目前都 wrap，但这是安全敏感路径（decode 恶意 RLP header），依赖未定义行为不如显式检查稳健。

## 三、Checked-and-clean（证伪的假设）

- **`parseQuantity` 与 `parseQuantityOrUInt64Number` 的 hex 路径是否不一致？** 否——两者 string 分支都 `safeFromQuantity`，行为一致。
- **构造器 ÷1000 无条件是否破坏 NON_ETH 头？** 否——「内部统一 ms」约定下，所有内部 header（含 NON_ETH 原生块、eth-genesis）都是 ms，÷1000 统一转秒，正确。
- **`applyEthGenesisHeader` 的 ×1000 与 morebtcg 的 overflow 点是否确为 UB？** 是——`m_timestamp` 是 int64_t（GenesisConfig.h:81），NodeConfig.cpp:370 只 bound 到 `u256(INT64_MAX)`，`m_timestamp*1000` 在 `≥ INT64_MAX/1000` 时 signed overflow（UB）。morebtcg 建议「NodeConfig bound 改 INT64_MAX/1000」是最便宜的正确修复。此条已在 morebtcg inline 覆盖，不重复挂行内。

## 四、验证边界

基于 head `816039a` 源码逐行核对（EthBlockHeader.cpp 全文、Ledger.cpp 的 applyEthGenesisHeader、EngineHelper.cpp 的 parseQuantity/parseQuantityOrUInt64Number、GenesisConfig.h、NodeConfig.cpp 的 bound）。**未本地跑 build/test**；op-node 实机验证与 4 个测试套件全绿采信 PR 描述。

## 五、下一步

无阻塞。建议按 morebtcg 的 3 条 + 本 Finding 1 做 follow-up 加固（尤其 NodeConfig bound 那条是 UB）。是否 approve 由你决定。

**Review by Shichen-Wu — APPROVED** [review]
